docs/src/constraints.md

```diff
@@ -1,5 +1,111 @@
+*less-than-or-equal-to* constraint, we consider this constraint to be a  *scalar
+affine -in- less than* constraint. More generally, we use the shorthand
+*function-in-set* to refer to constraints composed of different types of
+functions and sets. In the rest of this page, we will introduce the different
+types of functions and sets that JuMP knows about as needed. You can also more
+details about this *function-in-set* concept in the MathOptInterface
+documentation.
+
+## The `@constraint` macro
+
+Constraints are added to a JuMP model using the [`@constraint`](@ref) macro. It
+is similar to the [`@variable`](@ref) macro. Here is an example of how to add
+the constraint `2x <= 1` to a JuMP model:
+```jldoctest con1; setup = :(model = JuMP.Model(); @variable(model, x))
+julia> @constraint(model, con, 2x <= 1)
+con : 2 x <= 1.0
+```
+Wasn't that easy! Let's unpack what happened, because just like
+[`@variable`](@ref) there are a few subtle things going on.
+ 1. The mathematical constraint $2x \le 1$ was added to the model.
+ 2. A JuMP called `con` was created that acts as a reference to the
+    constraint.
+ 3. JuMP set the *name* of the constraint to `"con"`.
+
+Just like the Julia variables created in [`@variable`](@ref), `con` can be bound
+to a different value. For example:
+```jldoctest con1
+julia> con
+con : 2 x <= 1.0
+
+julia> con = 1
+1
+
+julia> con
+1
+```
+However, the reference can be retrieved by querying the model using the symbolic
+name:
+```jldoctest con1
+julia> con = model[:con]
+con : 2 x <= 1.0
+
+julia> con
+con : 2 x <= 1.0
+```
+Because the named variables and constraints are stored in the same namespace,
+creating a constraint with the same name as a variable or an existing constraint
+will result in an error. To overcome this limitation, it is possible to create
+anonymous constraints, just like it is possible to create
+[Anonymous JuMP variables](@ref). This is done by dropping the second argument
+to [`@constraint`](@ref):
+```jldoctest con1
+julia> con = @constraint(model, 2x <= 1)
+2 x <= 1.0
+```
+
+It is also possible use different comparison operators (e.g., `>=` and `==`) to
+create the following types of constraints:
+```jldoctest con1
+julia> @constraint(model, 2x >= 1)
+2 x >= 1.0
+
+julia> @constraint(model, 2x == 1)
+2 x == 1.0
+
+julia> @constraint(model, 1 <= 2x <= 3)
+2 x ∈ [1.0, 3.0]
+```
+
+## Constraint containers
+
+So far, we've adding constraints one-by-one. However, just like
+[Variable containers](@ref), JuMP provides a mechanism for building groups of
+constraints in *containers*. Three types of constraint containers are supported:
+`Array`s, `JuMPArray`s, and `Dict`ionaries. We explain each of these in the
+following.
+
+### Arrays
+
+```jldoctest constraint_arrays; setup=:(model=Model(); @variable(model, x))
+julia> @constraint(model, con[i = 1:2], i * x <= i + 1)
+2-element Array{ConstraintRef{Model,C,Shape} where Shape<:JuMP.AbstractShape where C,1}:
```

Comment:
I'm not sure about overloading `show` for specializations of built-in types like Arrays. It might break expectations or make the types seem more magical than they are.

Reply:
@blegat Do we know the shape at macro time so that we can fill in the `Shape` parameter?

Reply:
In most of the case yes. If someone does `@constraint(model, c[i=1:2], f[i] in s[i])` we cannot infer `C` nor `Shape` though. Even with `@constraint(model, f .== g)` we could say that the shape is `ScalarShape` and `C` is `MOI.ConstraintIndex{???, MOI.EqualTo{Float64})` so we are missing information. We might need to do some magic trick like what base does for determining the type of the output container when doing `[... for ... in ...]` or with broadcast and `map`. But that will be a compile time or run time, not macro time/parse time.